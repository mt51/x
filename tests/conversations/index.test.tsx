/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import mountTest from '../shared/mountTest';
import rtlTest from '../shared/rtlTest';
import { fireEvent, render } from '../utils';
import Conversations, { type Conversation } from '../../src/conversations';

const items: Conversation[] = [
  {
    key: 'demo1',
    label: 'What is Ant Design X ?',
    icon: <div id="conversation-test-id">icon</div>,
    group: 'pinned',
  },
  {
    key: 'demo2',
    label: (
      <div>
        Getting Started:{' '}
        <a
          target="_blank"
          href="https://ant-design.antgroup.com/index-cn"
          rel="noreferrer"
        >
          Ant Design !
        </a>
      </div>
    ),
  },
  {
    key: 'demo4',
    label: 'In Docker, use üêë Ollama and initialize',
  },
  {
    key: 'demo5',
    label: 'Expired, please go to the recycle bin to check',
    disabled: true,
  },
  // Used to test the case where there is no key
  {
    label: 'No key',
    disabled: true,
  } as any,
];

const menu = jest.fn().mockReturnValue({
  items: [
    {
      label: 'ÈáçÂëΩÂêç',
      key: 'mod',
    },
    {
      label: 'Âà†Èô§',
      key: 'delete',
      danger: true,
    },
  ],
});

const menuWithTriggerOfFunction = jest.fn().mockReturnValue({
  trigger: <div onClick={() => {}}>menuTriggerForFunctionButton</div>,
});

const menuWithTriggerOfReactNode = jest.fn().mockReturnValue({
  trigger: <div onClick={() => {}}>menuTriggerForReactNodeButton</div>,
});

describe('Conversations Component', () => {
  mountTest(() => <Conversations />);

  rtlTest(() => <Conversations items={items} groupable menu={menu} />);

  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('Conversations component work', () => {
    const { container } = render(<Conversations items={items} />);
    const element =
      container.querySelector<HTMLUListElement>('.ant-conversations');
    expect(element).toBeTruthy();
    expect(element).toMatchSnapshot();
  });

  it('should handle defaultActiveKey', () => {
    const { getByText } = render(
      <Conversations items={items} defaultActiveKey="demo1" />,
    );
    const activeItem = getByText('What is Ant Design X ?');
    expect(activeItem.parentNode).toHaveClass('ant-conversations-item-active');
  });

  it('should handle activeKey', () => {
    const { getByText } = render(
      <Conversations items={items} activeKey="demo1" />,
    );
    const activeItem = getByText('What is Ant Design X ?');
    expect(activeItem.parentNode).toHaveClass('ant-conversations-item-active');
  });

  it('should trigger onActiveChange', () => {
    const onActiveChange = jest.fn();
    const { getByText } = render(
      <Conversations items={items} onActiveChange={onActiveChange} />,
    );
    fireEvent.click(getByText('What is Ant Design X ?'));
    expect(onActiveChange).toHaveBeenCalledWith('demo1');
    fireEvent.click(getByText('In Docker, use üêë Ollama and initialize'));
    expect(onActiveChange).toHaveBeenCalledWith('demo4');
    fireEvent.click(
      getByText('Expired, please go to the recycle bin to check'),
    );
    expect(onActiveChange).toHaveBeenCalledWith('demo4');
  });

  it('should handle menu function', () => {
    const { getByText, container } = render(
      <Conversations items={items} menu={menu} defaultActiveKey="demo1" />,
    );
    expect(menu).toHaveBeenCalled();
    const menuElement = container.querySelector('.ant-conversations-menu-icon');
    expect(menuElement).toBeInTheDocument();
    fireEvent.click(menuElement as HTMLElement);
    expect(getByText('ÈáçÂëΩÂêç')).toBeInTheDocument();
    expect(getByText('Âà†Èô§')).toBeInTheDocument();
    fireEvent.click(menuElement as HTMLElement);
    const element = container.querySelector('.ant-dropdown-open');
    expect(element).not.toBeInTheDocument();
  });
  describe('should handle menu trigger function', () => {
    it('render node', async () => {
      const { findAllByText, container } = render(
        <Conversations
          items={items}
          menu={menuWithTriggerOfReactNode}
          defaultActiveKey="demo1"
        />,
      );
      expect(
        (await findAllByText('menuTriggerForReactNodeButton')).length,
      ).toBeGreaterThan(0);

      // default icon has't render
      const menuElement = container.querySelector(
        '.ant-conversations-menu-icon',
      );
      expect(menuElement).toBeNull();
    });

    it('render function node', async () => {
      const { findAllByText, container } = render(
        <Conversations
          items={items}
          menu={menuWithTriggerOfFunction}
          defaultActiveKey="demo1"
        />,
      );
      expect(
        (await findAllByText('menuTriggerForFunctionButton')).length,
      ).toBeGreaterThan(0);
      // default icon has't render
      const menuElement = container.querySelector(
        '.ant-conversations-menu-icon',
      );
      expect(menuElement).toBeNull();
    });
  });

  it('should group items when groupable is true', () => {
    const { getByText } = render(<Conversations items={items} groupable />);
    expect(getByText('pinned')).toBeInTheDocument();
  });

  it('should use custom group title component', () => {
    const { getByText } = render(
      <Conversations
        items={items}
        groupable={{ title: (group) => <div>{group}</div> }}
      />,
    );
    expect(getByText('pinned')).toBeInTheDocument();
  });

  it('should sort groups when groupable.sort is provided', () => {
    const sort = jest.fn().mockReturnValue(0);
    render(<Conversations items={items} groupable={{ sort }} />);
    expect(sort).toHaveBeenCalled();
  });

  it('should not group items when groupable is false', () => {
    const { queryByText } = render(
      <Conversations items={items} groupable={false} />,
    );
    expect(queryByText('pinned')).not.toBeInTheDocument();
  });
});
