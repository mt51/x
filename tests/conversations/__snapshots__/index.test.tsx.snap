// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conversations Component Conversations component work 1`] = `
<ul
  class="ant-conversations"
>
  <li
    class="ant-conversations-item"
  >
    <div
      class="ant-conversations-icon"
    >
      <div
        id="conversation-test-id"
      >
        icon
      </div>
    </div>
    <span
      aria-label="What is Ant Design X ?"
      class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-conversations-label"
    >
      What is Ant Design X ?
      <span
        aria-hidden="true"
        style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; word-break: keep-all; white-space: nowrap;"
      >
        lg
      </span>
      <span
        aria-hidden="true"
        style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
      >
        <span
          aria-hidden="true"
        >
          ...
        </span>
      </span>
    </span>
  </li>
  <li
    class="ant-conversations-item"
  >
    <span
      class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-conversations-label"
    >
      <div>
        Getting Started:
        <a
          href="https://ant-design.antgroup.com/index-cn"
          rel="noreferrer"
          target="_blank"
        >
          Ant Design !
        </a>
      </div>
      <span
        aria-hidden="true"
        style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; word-break: keep-all; white-space: nowrap;"
      >
        lg
      </span>
      <span
        aria-hidden="true"
        style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
      >
        <span
          aria-hidden="true"
        >
          ...
        </span>
      </span>
    </span>
  </li>
  <li
    class="ant-conversations-item"
  >
    <span
      aria-label="In Docker, use üêë Ollama and initialize"
      class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-conversations-label"
    >
      In Docker, use üêë Ollama and initialize
      <span
        aria-hidden="true"
        style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; word-break: keep-all; white-space: nowrap;"
      >
        lg
      </span>
      <span
        aria-hidden="true"
        style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
      >
        <span
          aria-hidden="true"
        >
          ...
        </span>
      </span>
    </span>
  </li>
  <li
    class="ant-conversations-item ant-conversations-item-disabled"
  >
    <span
      aria-label="Expired, please go to the recycle bin to check"
      class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-conversations-label"
    >
      Expired, please go to the recycle bin to check
      <span
        aria-hidden="true"
        style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; word-break: keep-all; white-space: nowrap;"
      >
        lg
      </span>
      <span
        aria-hidden="true"
        style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
      >
        <span
          aria-hidden="true"
        >
          ...
        </span>
      </span>
    </span>
  </li>
  <li
    class="ant-conversations-item ant-conversations-item-disabled"
  >
    <span
      aria-label="No key"
      class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-conversations-label"
    >
      No key
      <span
        aria-hidden="true"
        style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; word-break: keep-all; white-space: nowrap;"
      >
        lg
      </span>
      <span
        aria-hidden="true"
        style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
      >
        <span
          aria-hidden="true"
        >
          ...
        </span>
      </span>
    </span>
  </li>
</ul>
`;

exports[`Conversations Component rtl render component should be rendered correctly in RTL direction 1`] = `
<ul
  class="ant-conversations ant-conversations-rtl"
>
  <li>
    <div
      class="ant-conversations-group-title"
    >
      <span
        class="ant-typography ant-typography-rtl"
        direction="rtl"
      >
        pinned
      </span>
    </div>
    <ul
      class="ant-conversations-list"
    >
      <li
        class="ant-conversations-item"
      >
        <div
          class="ant-conversations-icon"
        >
          <div
            id="conversation-test-id"
          >
            icon
          </div>
        </div>
        <span
          aria-label="What is Ant Design X ?"
          class="ant-typography ant-typography-rtl ant-typography-ellipsis ant-typography-single-line ant-conversations-label"
          direction="rtl"
        >
          What is Ant Design X ?
          <span
            aria-hidden="true"
            style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; word-break: keep-all; white-space: nowrap;"
          >
            lg
          </span>
          <span
            aria-hidden="true"
            style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
          >
            <span
              aria-hidden="true"
            >
              ...
            </span>
          </span>
        </span>
        <span
          aria-label="ellipsis"
          class="anticon anticon-ellipsis ant-dropdown-trigger ant-dropdown-rtl ant-conversations-menu-icon"
          role="img"
          tabindex="-1"
        >
          <svg
            aria-hidden="true"
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </li>
    </ul>
  </li>
  <li>
    <div
      class="ant-conversations-group-title"
    />
    <ul
      class="ant-conversations-list"
    >
      <li
        class="ant-conversations-item"
      >
        <span
          class="ant-typography ant-typography-rtl ant-typography-ellipsis ant-typography-single-line ant-conversations-label"
          direction="rtl"
        >
          <div>
            Getting Started:
            <a
              href="https://ant-design.antgroup.com/index-cn"
              rel="noreferrer"
              target="_blank"
            >
              Ant Design !
            </a>
          </div>
          <span
            aria-hidden="true"
            style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; word-break: keep-all; white-space: nowrap;"
          >
            lg
          </span>
          <span
            aria-hidden="true"
            style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
          >
            <span
              aria-hidden="true"
            >
              ...
            </span>
          </span>
        </span>
        <span
          aria-label="ellipsis"
          class="anticon anticon-ellipsis ant-dropdown-trigger ant-dropdown-rtl ant-conversations-menu-icon"
          role="img"
          tabindex="-1"
        >
          <svg
            aria-hidden="true"
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </li>
      <li
        class="ant-conversations-item"
      >
        <span
          aria-label="In Docker, use üêë Ollama and initialize"
          class="ant-typography ant-typography-rtl ant-typography-ellipsis ant-typography-single-line ant-conversations-label"
          direction="rtl"
        >
          In Docker, use üêë Ollama and initialize
          <span
            aria-hidden="true"
            style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; word-break: keep-all; white-space: nowrap;"
          >
            lg
          </span>
          <span
            aria-hidden="true"
            style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
          >
            <span
              aria-hidden="true"
            >
              ...
            </span>
          </span>
        </span>
        <span
          aria-label="ellipsis"
          class="anticon anticon-ellipsis ant-dropdown-trigger ant-dropdown-rtl ant-conversations-menu-icon"
          role="img"
          tabindex="-1"
        >
          <svg
            aria-hidden="true"
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </li>
      <li
        class="ant-conversations-item ant-conversations-item-disabled"
      >
        <span
          aria-label="Expired, please go to the recycle bin to check"
          class="ant-typography ant-typography-rtl ant-typography-ellipsis ant-typography-single-line ant-conversations-label"
          direction="rtl"
        >
          Expired, please go to the recycle bin to check
          <span
            aria-hidden="true"
            style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; word-break: keep-all; white-space: nowrap;"
          >
            lg
          </span>
          <span
            aria-hidden="true"
            style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
          >
            <span
              aria-hidden="true"
            >
              ...
            </span>
          </span>
        </span>
      </li>
      <li
        class="ant-conversations-item ant-conversations-item-disabled"
      >
        <span
          aria-label="No key"
          class="ant-typography ant-typography-rtl ant-typography-ellipsis ant-typography-single-line ant-conversations-label"
          direction="rtl"
        >
          No key
          <span
            aria-hidden="true"
            style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; word-break: keep-all; white-space: nowrap;"
          >
            lg
          </span>
          <span
            aria-hidden="true"
            style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
          >
            <span
              aria-hidden="true"
            >
              ...
            </span>
          </span>
        </span>
      </li>
    </ul>
  </li>
</ul>
`;
